From 632881396a7221ec8a4f174c9ff56eb945bbf008 Mon Sep 17 00:00:00 2001
From: Andras Lasso <lasso@queensu.ca>
Date: Mon, 5 Oct 2026 02:39:55 -0400
Subject: [PATCH] BUG: Update segmentation views when a display node is
 modified during a batch

When a segmentation display node was modified while modified events of the
segmentation node were blocked (for example by MRMLNodeModifyBlocker, as the
Show 3D button does), the segmentation node invoked DisplayModifiedEvent
only once, at the end of the batch, without the display node as call data.
The 2D and 3D segmentation displayable managers ignored the event in this
case, so the views were not updated until another event came. For example,
choosing "Binary labelmap" as representation shown in 3D did not show the
segmentation in 3D views until a segment was added.

Now all display nodes of the segmentation are updated if the event does not
specify which display node was modified.
---
 .../MRMLDM/vtkMRMLSegmentationsDisplayableManager2D.cxx   | 8 +++++++-
 .../MRMLDM/vtkMRMLSegmentationsDisplayableManager3D.cxx   | 8 +++++++-
 2 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/Modules/Loadable/Segmentations/MRMLDM/vtkMRMLSegmentationsDisplayableManager2D.cxx b/Modules/Loadable/Segmentations/MRMLDM/vtkMRMLSegmentationsDisplayableManager2D.cxx
index 4ab80cd0eb..713d0946ef 100644
--- a/Modules/Loadable/Segmentations/MRMLDM/vtkMRMLSegmentationsDisplayableManager2D.cxx
+++ b/Modules/Loadable/Segmentations/MRMLDM/vtkMRMLSegmentationsDisplayableManager2D.cxx
@@ -1383,8 +1383,14 @@ void vtkMRMLSegmentationsDisplayableManager2D::ProcessMRMLNodesEvents(vtkObject*
       if (displayNode)
       {
         this->Internal->UpdateDisplayNode(displayNode);
-        this->RequestRender();
       }
+      else
+      {
+        // The event does not tell which display node was modified if it was invoked after modification
+        // of the segmentation node ended (MRMLNodeModifyBlocker): update all display nodes
+        this->Internal->UpdateAllDisplayNodesForSegment(displayableNode);
+      }
+      this->RequestRender();
     }
     else if ((event == vtkMRMLDisplayableNode::TransformModifiedEvent)      //
              || (event == vtkMRMLTransformableNode::TransformModifiedEvent) //
diff --git a/Modules/Loadable/Segmentations/MRMLDM/vtkMRMLSegmentationsDisplayableManager3D.cxx b/Modules/Loadable/Segmentations/MRMLDM/vtkMRMLSegmentationsDisplayableManager3D.cxx
index 8b413a7942..6eedc2a8bc 100644
--- a/Modules/Loadable/Segmentations/MRMLDM/vtkMRMLSegmentationsDisplayableManager3D.cxx
+++ b/Modules/Loadable/Segmentations/MRMLDM/vtkMRMLSegmentationsDisplayableManager3D.cxx
@@ -1349,8 +1349,14 @@ void vtkMRMLSegmentationsDisplayableManager3D::ProcessMRMLNodesEvents(vtkObject*
       if (displayNode)
       {
         this->Internal->UpdateDisplayNode(displayNode);
-        this->RequestRender();
       }
+      else
+      {
+        // The event does not tell which display node was modified if it was invoked after modification
+        // of the segmentation node ended (MRMLNodeModifyBlocker): update all display nodes
+        this->Internal->UpdateAllDisplayNodesForSegment(displayableNode);
+      }
+      this->RequestRender();
     }
     else if ((event == vtkMRMLDisplayableNode::TransformModifiedEvent)      //
              || (event == vtkMRMLTransformableNode::TransformModifiedEvent) //
-- 
2.44.0.windows.1

